tests/integration-tests-topologies/src/main/java/org/apache/pulsar/tests/topologies/PulsarCluster.java

```diff
@@ -136,52 +173,38 @@ public void start() throws Exception {
             "bin/init-cluster.sh");
         log.info("Successfully initialized the cluster.");
 
-        // create bookies
-        bookieContainers.putAll(
-            runNumContainers("bookie", spec.numBookies(), (name) -> new BKContainer(clusterName, name)
-                .withNetwork(network)
-                .withNetworkAliases(name)
-                .withEnv("zkServers", ZKContainer.NAME)
-                .withEnv("useHostNameAsBookieID", "true")
-                .withEnv("clusterName", clusterName)
-            )
-        );
+        // start bookies
+        bookieContainers.values().stream().forEach(BKContainer::start);
+        log.info("Successfully started {} bookie conntainers.", bookieContainers.values().size());
 
-        // create brokers
-        brokerContainers.putAll(
-            runNumContainers("broker", spec.numBrokers(), (name) -> new BrokerContainer(clusterName, name)
-                .withNetwork(network)
-                .withNetworkAliases(name)
-                .withEnv("zkServers", ZKContainer.NAME)
-                .withEnv("zookeeperServers", ZKContainer.NAME)
-                .withEnv("configurationStoreServers", CSContainer.NAME + ":" + CS_PORT)
-                .withEnv("clusterName", clusterName)
-                .withEnv("brokerServiceCompactionMonitorIntervalInSeconds", "1")
-            )
-        );
+        // start brokers
+        brokerContainers.values().stream().forEach(BrokerContainer::start);
+        log.info("Successfully started {} broker conntainers.", brokerContainers.values().size());
 
         // create proxy
         proxyContainer.start();
         log.info("Successfully started pulsar proxy.");
 
+        if(spec.enableTieredStorage()){
```

Comment:
spacing

Reply:
instead of checking if spec has it enabled, check if s3Container is not null. Or better, make s3Container an Optional and run stuff against it using ifPresent(x -> x.doStuff);

This way spec is only passed to constructor, you don't need to keep it around.

Reply:
will change to null check for now